src/main/python/ayab/engine/engine.py

```diff
@@ -146,13 +151,13 @@ def knit_config(self, image):
         # send signal to start knitting
         self.emit_knitting_starter()
 
-    def validate(self):
+    def validate(self)->tuple[Literal[False],str]|tuple[Literal[True],None]:
```

Comment:
"this tuple only has t[1] as a string if t[0] is False" is theoretically a really nice and tight typing, but, well, see previous comment.